src/Data/Conduit/Tar.hs

```diff
@@ -270,3 +245,482 @@ withEntries :: MonadThrow m
+                withFileInfo inner
+        Just (ChunkHeader h) -> do
+            payloadsConduit .| (inner (fileInfoFromHeader h) <* sinkNull)
+            withFileInfo inner
+        Just x@(ChunkPayload offset _bs) -> do
+            leftover x
+            throwM $ UnexpectedPayload offset
+        Just (ChunkException e) -> throwM e
+
+
+-- | Take care of custom GNU tar format.
+handleGnuTarHeader :: MonadThrow m
+                   => Header
+                   -> ConduitM TarChunk o m ()
+handleGnuTarHeader h = do
+    case headerLinkIndicator h of
+        76 -> do
+            let pSize = headerPayloadSize h
+            -- guard against names that are too long in order to prevent a DoS attack on unbounded
+            -- file names
+            unless (0 < pSize && pSize <= 4096) $
+                throwM $
+                FileTypeError (headerPayloadOffset h) 'L' $ "Filepath is too long: " ++ show pSize
+            longFileNameBuilder <- payloadsConduit .| sinkBuilder
+            let longFileName = SL.toStrict . SL.init . toLazyByteString $ longFileNameBuilder
+            mcNext <- await
+            case mcNext of
+                Just (ChunkHeader nh) -> do
+                    unless (S.isPrefixOf (fromShort (headerFileNameSuffix nh)) longFileName) $
+                        throwM $
+                        FileTypeError (headerPayloadOffset nh) 'L' $
+                        "Long filename doesn't match the original."
+                    leftover
+                        (ChunkHeader $
+                         nh
+                         { headerFileNameSuffix = toShort longFileName
+                         , headerFileNamePrefix = SS.empty
+                         })
+                Just c@(ChunkPayload offset _) -> do
+                    leftover c
+                    throwM $ InvalidHeader offset
+                Just (ChunkException exc) -> throwM exc
+                Nothing -> throwM NoMoreHeaders
+        83 -> do
+            payloadsConduit .| sinkNull -- discard sparse files payload
+            -- TODO : Implement restoring of sparse files
+        _ -> return ()
+
+
+-- | Just like `withFileInfo`, but works directly on the stream of bytes.
+untar :: MonadThrow m
+      => (FileInfo -> ConduitM ByteString o m ())
+      -> ConduitM ByteString o m ()
+untar inner = untarChunks .| withFileInfo inner
+
+
+-- | Just like `untar`, except that each `FileInfo` handling function can produce a finalizing
+-- action, all of which will be executed after the whole tarball has been processed in the opposite
+-- order. Very useful with `restoreFile` and `restoreFileInto`, since they restore direcory
+-- modification timestamps only after files have been fully written to disk.
+untarWithFinalizers ::
+       (MonadThrow m, MonadIO m)
+    => (FileInfo -> ConduitM ByteString (IO ()) m ())
+    -> ConduitM ByteString c m ()
+untarWithFinalizers inner = do
+    finilizers <- untar inner .| foldlC (flip (:)) []
```

Comment:
I think we already discussed this @lehins, but... instead of collecting a list of actions, how about combining them into a single action by folding with `>>`?

Reply:
Yep we sure did. I've sent you a message in slack about a month ago ;)

> Your idea with `(untar inner .| foldC) >>= liftIO` was good until I got to supporting GHC 7.10 :slightly_smiling_face:
```Could not deduce (Monoid (IO ())) arising from a use of 'foldC'```

Reply:
I could of course add some CPP, if you'd still prefer to use folding with `>>`

Reply:
Ah, right. But that still doesn't mean we need to go for the much less efficient list version, the `>>` approach should still work (even if it's not as nifty as `foldC` 😦)

Reply:
CPP shouldn't be necessary to use the `>>` approach, it's not relying on the `Monoid` instance, only the `Monad` instance. (`*>` is of course also an option, using the `Applicative` instance, but it doesn't really matter.)

Reply:
You are absolutely right, I can still fold with `>>`, plus using lists is certainly less then ideal, I'll make the change.

Reply:
I was just thinking about ordering of these finalizers. Does the order matter? It seems like you're performing them in reverse right now, which would work fine for the `>>` case as well. If you want to perform them in original order, it would make sense to apply the codensity transform style here, e.g.:

```haskell
actions <- ... .| foldlC (\front action -> front . (action >>)) id
actions (return ())
```

Reply:
Yeah, we've talked about the order, it doesn't really matter, as long as all of the actions are executed after the pipeline has finished. Nice trick with codensity transform though, I really like it :) Here is what I changed it too, just in case:
```
    finilizers <- untar inner .| foldlC (>>) (return ())
    liftIO finilizers
```

Reply:
If the ordering doesn't matter, that's _much_ nicer, and more efficient than the codensity approach.